src/machine/machine_nrf52.go

```diff
@@ -0,0 +1,155 @@
+// +build nrf52
+
+package machine
+
+import (
+	"device/nrf"
+	"unsafe"
+)
+
+// InitADC initializes the registers needed for ADC.
+func InitADC() {
+	return // no specific setup on nrf52 machine.
+}
+
+// Configure configures an ADC pin to be able to read analog data.
+func (a ADC) Configure() {
+	return // no pin specific setup on nrf52 machine.
+}
+
+// Get returns the current value of a ADC pin in the range 0..0xffff.
+func (a ADC) Get() uint16 {
+	var pwmPin uint32
+	var value int16
+
+	switch a.Pin {
+	case 2:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput0
+
+	case 3:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput1
+
+	case 4:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput2
+
+	case 5:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput3
+
+	case 28:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput4
+
+	case 29:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput5
+
+	case 30:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput6
+
+	case 31:
+		pwmPin = nrf.SAADC_CH_PSELP_PSELP_AnalogInput7
+
+	default:
+		return 0
+	}
+
+	// Set 14-bit sample resolution.
+	nrf.SAADC.RESOLUTION = nrf.SAADC_RESOLUTION_VAL_12bit
+
+	// Enable ADC.
+	nrf.SAADC.ENABLE = (nrf.SAADC_ENABLE_ENABLE_Enabled << nrf.SAADC_ENABLE_ENABLE_Pos)
+	for i := 0; i < 8; i++ {
+		nrf.SAADC.CH[i].PSELN = nrf.SAADC_CH_PSELP_PSELP_NC
+		nrf.SAADC.CH[i].PSELP = nrf.SAADC_CH_PSELP_PSELP_NC
+	}
+
+	// Configure ADC.
+	nrf.SAADC.CH[0].CONFIG = ((nrf.SAADC_CH_CONFIG_RESP_Bypass << nrf.SAADC_CH_CONFIG_RESP_Pos) & nrf.SAADC_CH_CONFIG_RESP_Msk) |
+		((nrf.SAADC_CH_CONFIG_RESP_Bypass << nrf.SAADC_CH_CONFIG_RESN_Pos) & nrf.SAADC_CH_CONFIG_RESN_Msk) |
+		((nrf.SAADC_CH_CONFIG_GAIN_Gain1_5 << nrf.SAADC_CH_CONFIG_GAIN_Pos) & nrf.SAADC_CH_CONFIG_GAIN_Msk) |
+		((nrf.SAADC_CH_CONFIG_REFSEL_Internal << nrf.SAADC_CH_CONFIG_REFSEL_Pos) & nrf.SAADC_CH_CONFIG_REFSEL_Msk) |
+		((nrf.SAADC_CH_CONFIG_TACQ_3us << nrf.SAADC_CH_CONFIG_TACQ_Pos) & nrf.SAADC_CH_CONFIG_TACQ_Msk) |
+		((nrf.SAADC_CH_CONFIG_MODE_SE << nrf.SAADC_CH_CONFIG_MODE_Pos) & nrf.SAADC_CH_CONFIG_MODE_Msk)
+
+	// Set pin to read.
+	nrf.SAADC.CH[0].PSELN = nrf.RegValue(pwmPin)
+	nrf.SAADC.CH[0].PSELP = nrf.RegValue(pwmPin)
+
+	// Destination for sample result.
+	nrf.SAADC.RESULT.PTR = nrf.RegValue(uintptr(unsafe.Pointer(&value)))
+	nrf.SAADC.RESULT.MAXCNT = 1 // One sample
+
+	// Start tasks.
+	nrf.SAADC.TASKS_START = 1
+	for nrf.SAADC.EVENTS_STARTED == 0 {
+	}
+	nrf.SAADC.EVENTS_STARTED = 0x00
+
+	// Start the sample task.
+	nrf.SAADC.TASKS_SAMPLE = 1
+
+	// Wait until the sample task is done.
+	for nrf.SAADC.EVENTS_END == 0 {
+	}
+	nrf.SAADC.EVENTS_END = 0x00
+
+	// Stop the ADC
+	nrf.SAADC.TASKS_STOP = 1
+	for nrf.SAADC.EVENTS_STOPPED == 0 {
+	}
+	nrf.SAADC.EVENTS_STOPPED = 0
+
+	// Disable the ADC.
+	nrf.SAADC.ENABLE = (nrf.SAADC_ENABLE_ENABLE_Disabled << nrf.SAADC_ENABLE_ENABLE_Pos)
+
+	if value < 0 {
+		value = 0
+	}
+
+	// Return 16-bit result from 12-bit value.
+	return uint16(value << 4)
```

Comment:
Note that this would have been undefined behavior in C, as you're shifting more than the bit width (15 bits without the sign, and the value can be bigger than 15 bits after the shift). Luckily this is not C so it is well defined. That is, I haven't seen anything in the Go spec about shift overflow but I know it is well defined in TinyGo as most integers are treated as unsigned by LLVM unless it is told otherwise.

It might be slightly more obvious what is going on here when you cast first to 16 bits before shifting.

See e.g. [this tweet](https://twitter.com/againsthimself/status/984947438631956485) for a related oddity in C.